commit 875bfd5276f31d09e811d31fca638b9f4d1205e8
Author: Kees Cook <kees@kernel.org>
Date:   Fri Sep 2 13:02:26 2022 -0700

    fortify: Add KUnit test for FORTIFY_SOURCE internals
    
    Add lib/fortify_kunit.c KUnit test for checking the expected behavioral
    characteristics of FORTIFY_SOURCE internals.
    
    Cc: Nick Desaulniers <ndesaulniers@google.com>
    Cc: Nathan Chancellor <nathan@kernel.org>
    Cc: Tom Rix <trix@redhat.com>
    Cc: Andrew Morton <akpm@linux-foundation.org>
    Cc: Vlastimil Babka <vbabka@suse.cz>
    Cc: "Steven Rostedt (Google)" <rostedt@goodmis.org>
    Cc: Yury Norov <yury.norov@gmail.com>
    Cc: Masami Hiramatsu <mhiramat@kernel.org>
    Cc: Sander Vanheule <sander@svanheule.net>
    Cc: linux-hardening@vger.kernel.org
    Cc: llvm@lists.linux.dev
    Reviewed-by: David Gow <davidgow@google.com>
    Signed-off-by: Kees Cook <keescook@chromium.org>

diff --git a/lib/Kconfig.debug b/lib/Kconfig.debug
index 36455953d306..1f267c0ddffd 100644
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -2542,6 +2542,15 @@ config STACKINIT_KUNIT_TEST
 	  CONFIG_GCC_PLUGIN_STRUCTLEAK, CONFIG_GCC_PLUGIN_STRUCTLEAK_BYREF,
 	  or CONFIG_GCC_PLUGIN_STRUCTLEAK_BYREF_ALL.
 
+config FORTIFY_KUNIT_TEST
+	tristate "Test fortified str*() and mem*() function internals at runtime" if !KUNIT_ALL_TESTS
+	depends on KUNIT && FORTIFY_SOURCE
+	default KUNIT_ALL_TESTS
+	help
+	  Builds unit tests for checking internals of FORTIFY_SOURCE as used
+	  by the str*() and mem*() family of functions. For testing runtime
+	  traps of FORTIFY_SOURCE, see LKDTM's "FORTIFY_*" tests.
+
 config TEST_UDELAY
 	tristate "udelay test driver"
 	help